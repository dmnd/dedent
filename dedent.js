// @flow

export default function dedent(
  strings: string | Array<string>,
  ...values: Array<string>
) {
  // $FlowFixMe: Flow doesn't undestand .raw
  const raw = typeof strings === "string" ? [strings] : strings.raw;

  // first, perform interpolation
  let result = "";
  for (let i = 0; i < raw.length; i++) {
    result += raw[i]
      // join lines when there is a suppressed newline
      .replace(/\\\n[ \t]*/g, "")
      // handle escaped backticks
      .replace(/\\`/g, "`");

    if (i < values.length) {
      result += values[i];
    }
  }

  // now strip indentation
  const lines = result.split("\n");
  let mindent: number | null = null;
  lines.forEach(l => {
    let m = l.match(/^(\s+)\S+/);
    if (m) {
      let indent = m[1].length;
      if (!mindent) {
        // this is the first indented line
        mindent = indent;
      } else {
        mindent = Math.min(mindent, indent);
      }
    }
  });

  if (mindent !== null) {
    const m = mindent; // appease Flow
    result = lines.map(l => l[0] === " " ? l.slice(m) : l).join("\n");
  }

  if (result.startsWith('\n')) {
    result = result.substr(1);
  }
  return result.replace(/^((.|\n)*)\n[ \t]*?$/, '$1');
}
