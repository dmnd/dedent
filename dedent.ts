export default function dedent(literals: string): string;
export default function dedent(
  strings: TemplateStringsArray,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ...values: any[]
): string;
export default function dedent(
  strings: string | TemplateStringsArray,
  ...values: string[]
) {
  const raw = typeof strings === "string" ? [strings] : strings.raw;

  // first, perform interpolation
  let result = "";
  for (let i = 0; i < raw.length; i++) {
    result += raw[i]
      // handle escaped newlines, backticks, and interpolation characters
      .replace(/\\`/g, "`")
      .replace(/\\\n[ \t]*/g, "")
      .replace(/\\\$/g, "$")
      .replace(/\\{/g, "{");

    if (i < values.length) {
      result += values[i];
    }
  }

  // now strip indentation
  const lines = result.split("\n");
  let mindent: number | null = null;
  for (const l of lines) {
    const m = l.match(/^(\s+)\S+/);
    if (m) {
      const indent = m[1].length;
      if (!mindent) {
        // this is the first indented line
        mindent = indent;
      } else {
        mindent = Math.min(mindent, indent);
      }
    }
  }

  if (mindent !== null) {
    const m = mindent; // appease TypeScript
    result = lines
      // https://github.com/typescript-eslint/typescript-eslint/issues/7140
      // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
      .map((l) => (l[0] === " " || l[0] === "\t" ? l.slice(m) : l))
      .join("\n");
  }

  return (
    result
      // dedent eats leading and trailing whitespace too
      .trim()
      // handle escaped newlines at the end to ensure they don't get stripped too
      .replace(/\\n/g, "\n")
  );
}
