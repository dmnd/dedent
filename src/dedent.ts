import type { Dedent, DedentOptions } from "./types.js";

export type * from "./types.js";

const dedent: Dedent = createDedent({});

export default dedent;

function createDedent(options: DedentOptions) {
	dedent.withOptions = (newOptions: DedentOptions): Dedent =>
		createDedent({ ...options, ...newOptions });

	return dedent;

	function dedent(literals: string): string;
	function dedent(strings: TemplateStringsArray, ...values: unknown[]): string;
	function dedent(
		strings: TemplateStringsArray | string,
		...values: unknown[]
	) {
		const raw = typeof strings === "string" ? [strings] : strings.raw;
		const { escapeSpecialCharacters = Array.isArray(strings) } = options;

		// first, perform interpolation
		let result = "";
		for (let i = 0; i < raw.length; i++) {
			let next = raw[i];

			if (escapeSpecialCharacters) {
				// handle escaped newlines, backticks, and interpolation characters
				next = next
					.replace(/\\\n[ \t]*/g, "")
					.replace(/\\`/g, "`")
					.replace(/\\\$/g, "$")
					.replace(/\\\{/g, "{");
			}

			result += next;

			if (i < values.length) {
				// eslint-disable-next-line @typescript-eslint/restrict-plus-operands
				result += values[i];
			}
		}

		// now strip indentation
		const lines = result.split("\n");
		let mindent: null | number = null;
		for (const l of lines) {
			const m = l.match(/^(\s+)\S+/);
			if (m) {
				const indent = m[1].length;
				if (!mindent) {
					// this is the first indented line
					mindent = indent;
				} else {
					mindent = Math.min(mindent, indent);
				}
			}
		}

		if (mindent !== null) {
			const m = mindent; // appease TypeScript
			result = lines
				// https://github.com/typescript-eslint/typescript-eslint/issues/7140
				// eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
				.map((l) => (l[0] === " " || l[0] === "\t" ? l.slice(m) : l))
				.join("\n");
		}

		return (
			result
				// dedent eats leading and trailing whitespace too
				.trim()
				// handle escaped newlines at the end to ensure they don't get stripped too
				.replace(/\\n/g, "\n")
		);
	}
}
