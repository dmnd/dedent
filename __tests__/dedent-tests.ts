import dedent from "../dedent";

describe("dedent", () => {
  it("works without interpolation", () => {
    expect(
      dedent`first
         second
         third`
    ).toMatchSnapshot();
  });

  it("works with interpolation", () => {
    expect(
      dedent`first ${"line"}
         ${"second"}
         third`
    ).toMatchSnapshot();
  });

  it("works with suppressed newlines", () => {
    expect(
      dedent`first \
         ${"second"}
         third`
    ).toMatchSnapshot();
  });

  it("works with blank first line", () => {
    expect(dedent`
      Some text that I might want to indent:
        * reasons
        * fun
      That's all.
    `).toMatchSnapshot();
  });

  it("works with multiple blank first lines", () => {
    expect(
      dedent`

         first
         second
         third`
    ).toMatchSnapshot();
  });

  it("works with removing same number of spaces", () => {
    expect(
      dedent`
         first
            second
               third
      `
    ).toMatchSnapshot();
  });

  describe("single line input", () => {
    it("works with single line input", () => {
      expect(dedent`A single line of input.`).toMatchSnapshot();
    });

    it("works with single line and closing backtick on newline", () => {
      expect(dedent`
        A single line of input.
      `).toMatchSnapshot();
    });

    it("works with single line and inline closing backtick", () => {
      expect(dedent`
        A single line of input.`).toMatchSnapshot();
    });
  });

  it("can be used as a function", () => {
    expect(
      dedent(`
      A test argument.
    `)
    ).toMatchSnapshot();
  });

  describe("function character escapes", () => {
    describe("default behavior", () => {
      it("does not escape backticks", () => {
        expect(dedent(`\``)).toMatchSnapshot();
      });

      it("does not escape dollar signs", () => {
        expect(dedent(`$`)).toMatchSnapshot();
      });

      it("does not escape opening braces", () => {
        expect(dedent(`{`)).toMatchSnapshot();
      });

      it("escapes double-escaped backticks", () => {
        expect(dedent(`\\\``)).toMatchSnapshot();
      });

      it("escapes double-escaped dollar signs", () => {
        expect(dedent(`\\$`)).toMatchSnapshot();
      });

      it("escapes double-escaped opening braces", () => {
        expect(dedent(`\\{`)).toMatchSnapshot();
      });

      it("ignores closing braces", () => {
        expect(dedent(`}`)).toMatchSnapshot();
      });
    });

    describe.each([undefined, false, true])(
      "with escapeSpecialCharacters %s",
      (escapeSpecialCharacters) => {
        test("backticks", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`\``)
          ).toMatchSnapshot();
        });

        test("dollar signs", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`$`)
          ).toMatchSnapshot();
        });

        test("opening braces", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`{`)
          ).toMatchSnapshot();
        });

        test("double-escaped backticks", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`\\\``)
          ).toMatchSnapshot();
        });

        test("double-escaped dollar signs", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`\\$`)
          ).toMatchSnapshot();
        });

        test("double-escaped opening braces", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })(`\\{`)
          ).toMatchSnapshot();
        });
      }
    );
  });

  describe("string tag character escapes", () => {
    describe("default behavior", () => {
      it("escapes backticks", () => {
        expect(dedent`\``).toMatchSnapshot();
      });

      it("escapes dollar signs", () => {
        expect(dedent`\$`).toMatchSnapshot();
      });

      it("escapes opening braces", () => {
        expect(dedent`\{`).toMatchSnapshot();
      });

      it("ignores closing braces", () => {
        expect(dedent`\}`).toMatchSnapshot();
      });
    });

    describe.each([undefined, false, true])(
      "with escapeSpecialCharacters %s",
      (escapeSpecialCharacters) => {
        test("backticks", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })`\``
          ).toMatchSnapshot();
        });

        test("dollar signs", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })`\$`
          ).toMatchSnapshot();
        });

        test("opening braces", () => {
          expect(
            dedent.options({ escapeSpecialCharacters })`\{`
          ).toMatchSnapshot();
        });
      }
    );
  });

  it("doesn't strip explicit newlines", () => {
    expect(dedent`
      <p>Hello world!</p>\n
    `).toMatchSnapshot();
  });

  it("doesn't strip explicit newlines with mindent", () => {
    expect(dedent`
      <p>
        Hello world!
      </p>\n
    `).toMatchSnapshot();
  });

  /* eslint-disable indent */
  it("works with tabs for indentation", () => {
    expect(
      dedent`
			first
				second
					third
			`
    ).toMatchSnapshot();
  });

  it("works with escaped tabs for indentation", () => {
    expect(dedent("\t\tfirst\n\t\t\tsecond\n\t\t\t\tthird")).toMatchSnapshot();
  });
  /* eslint-enable indent */
});
