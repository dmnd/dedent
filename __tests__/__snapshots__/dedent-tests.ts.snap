// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dedent can be used as a function 1`] = `"A test argument."`;

exports[`dedent doesn't strip exlicit newlines 1`] = `
"<p>Hello world!</p>
"
`;

exports[`dedent doesn't strip exlicit newlines with mindent 1`] = `
"<p>
  Hello world!
</p>
"
`;

exports[`dedent escapes backticks 1`] = `"\`"`;

exports[`dedent single line input works with single line and closing backtick on newline 1`] = `"A single line of input."`;

exports[`dedent single line input works with single line and inline closing backtick 1`] = `"A single line of input."`;

exports[`dedent single line input works with single line input 1`] = `"A single line of input."`;

exports[`dedent works with blank first line 1`] = `
"Some text that I might want to indent:
  * reasons
  * fun
That's all."
`;

exports[`dedent works with escaped tabs for indentation 1`] = `
"first
	second
		third"
`;

exports[`dedent works with interpolation 1`] = `
"first line
second
third"
`;

exports[`dedent works with multiple blank first lines 1`] = `
"first
second
third"
`;

exports[`dedent works with removing same number of spaces 1`] = `
"first
   second
      third"
`;

exports[`dedent works with suppressed newlines 1`] = `
"first second
third"
`;

exports[`dedent works with tabs for indentation 1`] = `
"first
	second
		third"
`;

exports[`dedent works without interpolation 1`] = `
"first
second
third"
`;
